keras_nlp/layers/preprocessing/mlm_mask_generator.py

```diff
@@ -0,0 +1,216 @@
+            between 0 and 1 which indicates how often a random token is
+            substituted for tokens selected for masking. Default is 0.1.
+            Note: mask_token_rate + random_token_rate <= 1.
+        padding_token_id: int, defaults to None. The id of padding token.
+
+    Input:
+        A 1D integer tensor of shape [sequence_length,] or a 2D integer tensor
+        of shape [batch_size, sequence_length], or a 2D integer RaggedTensor.
+        Represents the sequence to mask.
+
+    Returns:
+        A Dict with 4 keys:
+            tokens: Tensor, has the same type and shape of input.
+                Sequence after getting masked.
+            mask_positions: Tensor, or RaggedTensor if
+                `mask_selection_length` is None. The positions of tokens
+                getting masked.
+            mask_ids: Tensor, or RaggedTensor if
+                `mask_selection_length` is None. The original token ids
+                at masked positions.
+            mask_weights: Tensor, or RaggedTensor if
+                `mask_selection_length` is None. `mask_weights` has the same
+                shape as `mask_positions` and `mask_ids`. Each element in
+                `mask_weights` should be 0 or 1, 1 means the corresponding
+                position in `mask_positions` is an actual mask, 0 means it is
+                a pad.
+
+    Examples:
+
+    Basic usage.
+    >>> masker = keras_nlp.preprocessing.MLMMaskGenerator( \
+    ...     10, mask_selection_rate=0.2, mask_selection_length=5)
+    >>> masker(tf.constant([1, 2, 3, 4, 5]))
+    {'tokens': <tf.Tensor: shape=(5,), dtype=int32,
+        numpy=array([1, 2, 3, 4, 0], dtype=int32)>,
+    'mask_positions': <tf.Tensor: shape=(5,), dtype=int64,
+        numpy=array([4, 0, 0, 0, 0])>,
+    'mask_ids': <tf.Tensor: shape=(5,), dtype=int32,
+        numpy=array([5, 0, 0, 0, 0], dtype=int32)>,
+    'mask_weights': <tf.Tensor: shape=(1, 5), dtype=int64,
+        numpy=array([[1, 0, 0, 0, 0]])>}
+
+    Ragged Input:
+    >>> masker = keras_nlp.preprocessing.MLMMaskGenerator( \
+    ...     10, mask_selection_rate=0.5, mask_selection_length=5)
+    >>> masker(tf.ragged.constant([[1, 2], [1, 2, 3, 4]]))
+    {'tokens': <tf.RaggedTensor [[1, 4], [0, 2, 3, 0]]>,
+    'mask_positions': <tf.Tensor: shape=(2, 5), dtype=int64, numpy=
+        array([[1, 0, 0, 0, 0],
+               [0, 3, 0, 0, 0]])>,
+    'mask_ids': <tf.Tensor: shape=(2, 5), dtype=int32, numpy=
+        array([[2, 0, 0, 0, 0],
+               [1, 4, 0, 0, 0]], dtype=int32)>,
+    'mask_weights': <tf.Tensor: shape=(2, 5), dtype=int64, numpy=
+        array([[1, 0, 0, 0, 0],
+               [1, 1, 0, 0, 0]])>}
+    """
+
+    def __init__(
+        self,
+        vocabulary_size,
+        mask_selection_rate,
+        mask_selection_length=None,
+        unselectable_token_ids=None,
+        mask_token_id=0,
+        mask_token_rate=0.8,
+        random_token_rate=0.1,
+        padding_token_id=None,
```

Comment:
We should default padding token id to 0, that is the norm for most tokenizers.

Reply:
Actually even better, why not remove this, and default `unselectable_token_ids=[0]`. That covers the exact same use case, and cuts down an argument.

Reply:
Depending on the default value seems a bit error-prone to me, if the user overrides `unselectable_token_ids`  not including 0, then we will break our assumptions?

Reply:
Also it would change how masks are computed. suppose we have input [1, 2, 3, 4, 0, 0, 0, 0], setting `unselectable_token_ids=[0]` and `mask_selection_rate=0.5` will get 4 positions to mask, but if the selected position is 0, mask is skipped. This is to say, we may end up with as masking out 0 tokens to 4 tokens. If 0 token is masked out, then the example is not doing anything, if all 4 tokens are masked out, then this example is pure noise. 

I would prefer to keep this `padding_token_id` for converting dense tensor to RaggedTensor.

Reply:
Oh, for the default behavior of tf.text is to cap on number of selections before considering `unselectable_token_ids`?? If so that is a bad API.

If you say `unselectable_token_ids=[99], mask_selection_length=2 mask_selection_rate=0.5`, and then supply a array of `[99, 99, 99, 1, 99, 2, 99, 3, 99, 4, 99, 99, 99, 99, 99, 99]`, you should expect two mask positions on average not close to zero! Are you sure that is what tf.text is doing?

Re initial comment, if the user supplies `unselectable_token_ids` themselves, and they have a padded tensor with some padded token value, if seems perfectly reasonable to ask them to include that token value in there list. To me that is making less assumptions about the input, not more.

Reply:
Yes, from the code: https://github.com/tensorflow/text/blob/v2.8.1/tensorflow_text/python/ops/item_selector_ops.py#L238

I guess for now, we will default padding_token_id to 0, and default unselectable_token_ids to [0] to keep them in sync?

Reply:
The line you are linking is after a boolean_mask line. Check this out ->
https://colab.research.google.com/gist/mattdangerw/06a07a3f7b1fb7f0f179d04a848935b2/tf-text-ragged-test.ipynb

I don't think what you are describing as the behavior is true. We can remove `padding_token_id` and not run into the problem you describe.

Also think about what padding_token_id is really expressing--"An unmaskable token id, but only if your input is dense, and only for times this token appears consecutively at the end of your input. Anywhere else we see this (consecutively at the end of a ragged, or in the middle of a dense), we will consider it maskable." That's a confusing thing to attempt to explain to a user.

I think this argument hurts more than it adds.

Reply:
Ah, you are right. 

SGTM, will delete this argument.